File: src/services/responsePersonalizer.js
/**
 * Response Personalizer Service
 * Adapts AI responses based on customer profile and intent scores.
 */
class ResponsePersonalizer {
  constructor() {
    this.personalityMatrix = {
      young_professional: {
        language_style: 'casual_direct',
        payment_emphasis: 'convenience',
        urgency_level: 'medium',
        discount_approach: 'gradual_concessions',
      },
      middle_aged: {
        language_style: 'respectful_formal',
        payment_emphasis: 'family_responsibility',
        urgency_level: 'medium',
        discount_approach: 'fair_consideration',
      },
      senior_citizen: {
        language_style: 'patient_respectful',
        payment_emphasis: 'legacy_planning',
        urgency_level: 'low',
        discount_approach: 'generous_accommodation',
      },
      financial_stress: {
        language_style: 'empathetic_supportive',
        payment_emphasis: 'flexibility',
        urgency_level: 'very_low',
        discount_approach: 'significant_concessions',
      },
    };

    this.greetingTemplates = {
      young_professional: [
        'Oi {name}! Tudo bem? Vi que tem uma pendência de R$ {amount} no consórcio. Bora resolver isso rapidinho? Tenho umas opções que vão facilitar pra você!',
        'E aí {name}! Passando pra falar sobre sua parcela de R$ {amount}. Tenho umas condições especiais que acho que vão te interessar!',
      ],
      middle_aged: [
        'Olá {name}! Vi que você tem uma pendência no seu consórcio de R$ {amount}. Posso te ajudar a encontrar uma solução que funcione no seu bolso?',
        'Bom dia {name}! Vim conversar sobre sua pendência do consórcio para encontrarmos uma saída juntos. Temos algumas opções que podem te ajudar!',
      ],
      senior_citizen: [
        'Bom dia {name}! Com licença, gostaria de conversar sobre sua parcela do consórcio. Temos condições muito especiais que podem te ajudar a regularizar. Posso explicar?',
        'Olá {name}! Espero que esteja bem. Estou entrando em contato sobre seu consórcio. Gostaria de apresentar algumas opções flexíveis de pagamento, posso?',
      ],
      financial_stress: [
        'Olá {name}, tudo bem? Entendo que momentos difíceis acontecem. Estou aqui pra te ajudar a encontrar um caminho viável pra regularizar sua situação. Sem pressão, vamos conversar?',
        'Oi {name}! Sei que a situação pode estar complicada. Quero te ajudar a resolver isso da melhor forma possível. Temos opções bem flexíveis, posso explicar?',
      ],
    };
  }

  determineCluster(profile, intentScores) {
    const age = profile?.age || 35;
    const incomeLevel = profile?.income_level || profile?.income_bracket || 'middle';
    const financialCapacity = intentScores?.financial_capacity?.score || 0.5;

    if (financialCapacity < 0.3) return 'financial_stress';
    if (age < 35 && (incomeLevel === 'high' || incomeLevel === 'middle')) return 'young_professional';
    if (age >= 55) return 'senior_citizen';
    return 'middle_aged';
  }

  generateGreeting(customer, intentScores = {}) {
    const cluster = this.determineCluster(customer.profile, intentScores);
    const templates = this.greetingTemplates[cluster];
    const template = templates[Math.floor(Math.random() * templates.length)];

    const firstName = (customer.name || 'Cliente').split(' ')[0];
    const amount = (customer.overdue_amount || 0).toLocaleString('pt-BR', {
      minimumFractionDigits: 2,
    });

    return template.replace('{name}', firstName).replace('{amount}', amount);
  }

  generatePaymentOptions(customer) {
    const amount = customer.overdue_amount || 0;
    const cluster = this.determineCluster(customer.profile, {});
    const personality = this.personalityMatrix[cluster];
    const options = [];

    // Option 1: Full payment with discount
    let discountPct = 5;
    if (personality.discount_approach === 'generous_accommodation') discountPct = 15;
    else if (personality.discount_approach === 'significant_concessions') discountPct = 12;
    else if (personality.discount_approach === 'fair_consideration') discountPct = 8;

    const discountedAmount = amount * (1 - discountPct / 100);
    options.push({
      type: 'full_payment',
      description: `Pagamento à vista com ${discountPct}% de desconto`,
      amount: Math.round(discountedAmount * 100) / 100,
      discount_percentage: discountPct,
    });

    // Option 2: 3x installments
    const installment3 = Math.round((amount / 3) * 100) / 100;
    options.push({
      type: 'installment_3x',
      description: `Parcelamento em 3x de R$ ${installment3.toLocaleString('pt-BR', { minimumFractionDigits: 2 })}`,
      amount,
      installments: 3,
      installment_value: installment3,
    });

    // Option 3: 6x installments with small addition
    const total6 = amount * 1.05;
    const installment6 = Math.round((total6 / 6) * 100) / 100;
    options.push({
      type: 'installment_6x',
      description: `Parcelamento em 6x de R$ ${installment6.toLocaleString('pt-BR', { minimumFractionDigits: 2 })}`,
      amount: Math.round(total6 * 100) / 100,
      installments: 6,
      installment_value: installment6,
    });

    return options;
  }

  personalize(baseResponse, customer, intentScores) {
    const cluster = this.determineCluster(customer.profile, intentScores);
    const personality = this.personalityMatrix[cluster];
    return {
      message: baseResponse,
      personality_cluster: cluster,
      personality_settings: personality,
      payment_options: this.generatePaymentOptions(customer),
    };
  }
}

module.exports = new ResponsePersonalizer();

File: python/response_personalizer.py
"""
Response Personalizer - Adapts AI responses based on customer profiles.

Usage:
    python response_personalizer.py
"""
import json
import random
import sys


class ResponsePersonalizer:
    def __init__(self):
        self.personality_matrix = {
            "young_professional": {
                "language_style": "casual_direct",
                "payment_emphasis": "convenience",
                "urgency_level": "medium",
                "discount_approach": "gradual_concessions",
            },
            "middle_aged": {
                "language_style": "respectful_formal",
                "payment_emphasis": "family_responsibility",
                "urgency_level": "medium",
                "discount_approach": "fair_consideration",
            },
            "senior_citizen": {
                "language_style": "patient_respectful",
                "payment_emphasis": "legacy_planning",
                "urgency_level": "low",
                "discount_approach": "generous_accommodation",
            },
            "financial_stress": {
                "language_style": "empathetic_supportive",
                "payment_emphasis": "flexibility",
                "urgency_level": "very_low",
                "discount_approach": "significant_concessions",
            },
        }

        self.greeting_templates = {
            "young_professional": [
                "Oi {name}! Tudo bem? Vi que tem uma pendência de R$ {amount} no consórcio. Bora resolver isso rapidinho?",
                "E aí {name}! Passando pra falar sobre sua parcela de R$ {amount}. Tenho condições especiais!",
            ],
            "middle_aged": [
                "Olá {name}! Vi que você tem uma pendência no seu consórcio de R$ {amount}. Posso te ajudar?",
                "Bom dia {name}! Vim conversar sobre sua pendência do consórcio. Temos opções que podem te ajudar!",
            ],
            "senior_citizen": [
                "Bom dia {name}! Gostaria de conversar sobre sua parcela do consórcio. Temos condições especiais.",
                "Olá {name}! Espero que esteja bem. Gostaria de apresentar opções flexíveis de pagamento.",
            ],
            "financial_stress": [
                "Olá {name}, entendo que momentos difíceis acontecem. Estou aqui pra ajudar. Sem pressão.",
                "Oi {name}! Sei que a situação pode estar complicada. Temos opções bem flexíveis.",
            ],
        }

    def determine_cluster(self, profile, intent_scores=None):
        age = profile.get("age", 35)
        income_level = profile.get("income_level", profile.get("income_bracket", "middle"))
        financial_capacity = 0.5
        if intent_scores:
            financial_capacity = intent_scores.get("financial_capacity", 0.5)

        if financial_capacity < 0.3:
            return "financial_stress"
        if age < 35 and income_level in ("high", "middle"):
            return "young_professional"
        if age >= 55:
            return "senior_citizen"
        return "middle_aged"

    def personalize_response(self, base_response, customer_profile, intent_scores=None):
        cluster = self.determine_cluster(customer_profile, intent_scores)
        personality = self.personality_matrix[cluster]

        return {
            "message": base_response,
            "personality_cluster": cluster,
            "personality_settings": personality,
        }

    def generate_greeting(self, customer_name, overdue_amount, profile):
        cluster = self.determine_cluster(profile)
        templates = self.greeting_templates[cluster]
        template = random.choice(templates)
        first_name = customer_name.split()[0] if customer_name else "Cliente"
        amount = f"{overdue_amount:,.2f}".replace(",", "X").replace(".", ",").replace("X", ".")
        return template.format(name=first_name, amount=amount)


if __name__ == "__main__":
    personalizer = ResponsePersonalizer()

    # Demo with different profiles
    profiles = [
        {"name": "Lucas Mendes", "age": 28, "income_level": "high", "amount": 1500.00},
        {"name": "Roberto Silva", "age": 62, "income_level": "middle", "amount": 3200.50},
        {"name": "Ana Paula", "age": 40, "income_level": "low", "amount": 5000.00},
    ]

    for p in profiles:
        profile = {"age": p["age"], "income_level": p["income_level"]}
        cluster = personalizer.determine_cluster(profile)
        greeting = personalizer.generate_greeting(p["name"], p["amount"], profile)
        print(f"\nProfile: {p['name']} (age={p['age']}, income={p['income_level']})")
        print(f"Cluster: {cluster}")
        print(f"Greeting: {greeting}")

    print(json.dumps(personalizer.personality_matrix, indent=2, ensure_ascii=False))
